### ISSUE TITLE
[Analyzer] Topic Modeling

### ISSUE THREAD

**Issue by shahrukhx01** [obsei/obsei#131]
https://github.com/MaartenGr/BERTopic

**Comment by shahrukhx01** [comment]
@lalitpagaria you can assign me this and #130

**Comment by akar5h** [comment]
@shahrukhx01 Shahrukh pls confirm if you've started anything on this yet ? if not I would like to collaborate .

**Comment by shahrukhx01** [comment]
Hi Akarsh, I’ve not started anything yet. However before starting I’d suggest you to go through this:
https://www.sbert.netexamples/applications/clustering/README.html
this would also be helpful in #130

**Comment by akar5h** [comment]
thanks @shahrukhx01 , I went through the sentence_transformers libraries and models for clustering and have some idea myself. 
@lalitpagaria Hi , can you please elaborate more on Topic Modeling requirements . I assume our focus is on short texts , like reviews , is that right ?

**Comment by lalitpagaria** [comment]
@akar5h Actually main idea to perform do it on both short (reviews) and large text (news articles, emails etc) as well. But yes we could first start on small texts. This how I see it from user point of view -
1. User load data from DB/CSV/Excel/Airtable etc
2. Perform topic modeling on whole set using selected techniques (clustering algo, LDA, PCA)

This would be helpful for user to perform per-processing and gain taste about what all texts are talking about.
Feel free to correct if I wrote incorrectly in context of data science.

**Comment by shahrukhx01** [comment]
Just adding to Lalit's comment, when I was adding issue, my intention was to have a comparable pipeline to the zero-shot classifier, the end goal is to categorize data into categories/clusters with user-defined categories without any fine-tuning/training.

**Comment by lalitpagaria** [comment]
Yes @shahrukhx01 this is important and very helpful to user who doesn't have resource to run Obsei on GPUs. Even few Obsei users are asking it as well. So this will another classification analyzer.

**Comment by akar5h** [comment]
@shahrukhx01 @lalitpagaria 
Thanks for the input, from the discussion I will start Building a (Semi Supervised) Clustering pipeline to be used as another classification analyzer. 
https://github.com/MaartenGr/BERTopic like a good start for this . 

> Yes @shahrukhx01 this is important and very helpful to user who doesn't have resource to run Obsei on GPUs. Even few Obsei users are asking it as well. So this will another classification analyzer.
> Just adding to Lalit's comment, when I was adding issue, my intention was to have a comparable pipeline to the zero-shot classifier, the end goal is to categorize data into categories/clusters with user-defined categories without any fine-tuning/training.

Also, I feel `Topic Modeling` can have a separate module eventually, more practical to use it as a visualizer/preprocessor using some clustering algos .

**Comment by lalitpagaria** [comment]
Great @akar5h. Looking for ward for it :)

**Comment by lalitpagaria** [comment]
Adding one more https://github.com/ddangelov/Top2Vec

**Comment by akar5h** [comment]
Hi Lalit @lalitpagaria 
I have worked out first iteration of `Topic Analyzer`,  I want to clarify on output format from the `TopicAnalyzer`:
 
This topic analyzer takes in unlabelled texts as input and clusters them . It inherits from BaseAnalyzer so , has analyze_input as its base function, 
After clustering the texts , I calculate the most frequent texts in these clusters , and use them as the representation of the cluster. The output of analyze_input is List[TextPayload] where each TextPayload has this " representation of the cluster" as processed texts and meta: contains a list of all TextPayload that belong in this cluster , 
Example: Cluster1 most frequent Topics = ["Happy", "fast", delivery"] , CLuster 2 = ["worse","bad", "agent"] as  representation of the cluster, 
Then the output if analyzeinput will be `[TextPayload(processed_text = "Happy_fast_delivert", meta: [list of textpayloads in this cluster]), TextPayload(processed_text = "worse_bad_agent", meta: [list of textpayloads in this cluster])]`

So the above cluster and their representation of cluster is fine output format or you expect these calculated labels in some other format ? 

You can see this implemented here:
https://github.com/akar5h/obsei/blob/topic-analyzer/obsei/analyzer/topic_analyzer.py

Will create a PR when I complete a non deep learning based approach : "LDA"  integrated to this analyzer by end of the week .

**Comment by lalitpagaria** [comment]
Thanks a lot @akar5h 
Let me review it and provide you early review.

**Comment by lalitpagaria** [comment]
@akar5h I just had cursory looks and it looks fine to me. There are few code structure related things which we can take is forward on PR review.

@shahrukhx01 can you please have a look as you have more context in this field.

**Comment by shahrukhx01** [comment]
@lalitpagaria sure I will take a look at it over the weekend.

**Comment by lalitpagaria** [comment]
@akar5h could you please create PR so we can discuss on it.

**Comment by akar5h** [comment]
@lalitpagaria  , Will do surely , running on low bandwidth with office this week

**Comment by shahrukhx01** [comment]
Hi @lalitpagaria @akar5h I'm really sorry. I have been super consumed on couple of other things lately, I will try giving my input on this within this week.

**Comment by lalitpagaria** [comment]
@akar5h @shahrukhx01 No issue please take your time, no urgency :)
